### ISSUE TITLE
["Request"] Add a `fold` function for nullable types

### ISSUE THREAD

**Issue by LordRaydenMK** [arrow-kt/arrow-core#241]
What version are you currently using?

0.11

What would you like to see?

When moving from `Option` to using nullable types, the `fold` function is missing so you end up doing:

`?.let { ... } ?: run { ... }`

We can add a function:

```
fun <A, B> A?.fold(ifNull: () -> B, ifPresent: (A) -> B): B =
    this?.let(ifPresent) ?: ifNull()
```

that would make this a bit easier to work with.

Source: https://github.com/arrow-kt/arrow-core/pull/239#discussion_r495479486 suggested by @pablisco 

You can also see the pattern in the PR.

**Comment by pablisco** [comment]
If we mirror the current Option API for nullables it will make migration much easier as we can use ReplaceWith without any compilation errors after the replacement. Aditionally, having `map`, `flatMap` and such would be a nice companion to `stdlib`

**Comment by raulraja** [comment]
Can we make these inline?

**Comment by pablisco** [comment]
After more time than I would like to admit, I realised one problem...
`A?` is contravariant of all `Kind`ed types, so it adds ambiguity and, for obvious reasons, it cannot be resolved 🤦 

Maybe if we call it `foldNull`? I'm uncertain if there will be inline classes in the future. This indicates that could be (maybe?) https://github.com/Kotlin/KEEP/blob/master/proposals/inline-classes.md#inline-classes-over-nullable-reference-types

That would be the ideal solution, in reality, so we can avoid polluting the global scope with extension functions like these.

**Comment by nomisRev** [comment]
Yes, we currently also have this issue with `mapN` defined in `Nullable.kt` it's conflicing with `mapN` for `List` and possibly other top-level methods such as `mapN`, `tupledN` we might want to add for other types. We could define these functions on `object Nullable`, but that'd be a breaking change with the current API :(

See https://github.com/arrow-kt/arrow-core/pull/276#discussion_r546274480
